src/main/java/net/glowstone/entity/passive/GlowFirework.java

```diff
@@ -0,0 +1,204 @@
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
+import org.bukkit.event.entity.FireworkExplodeEvent;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.FireworkMeta;
+import org.bukkit.util.Vector;
+
+public class GlowFirework extends GlowEntity implements Firework {
+
+    @Getter
+    @Setter
+    private ItemStack item;
+    @Getter
+    @Setter
+    private UUID spawningEntity;
+    @Getter
+    private LivingEntity boostedEntity;
+    /**
+     * The number of ticks before this fireworks rocket explodes.
+     */
+    @Getter
+    @Setter
+    private int lifeTime;
+
+    public GlowFirework(Location location) {
+        super(location);
+        setSize(0.25f, 0.25f);
+    }
+
+    public GlowFirework(Location location, UUID spawningEntity, LivingEntity boostedEntity, ItemStack item) {
+        super(location);
+        this.spawningEntity = spawningEntity;
+        setBoostedEntity(boostedEntity);
+        setSize(0.25f, 0.25f);
+
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        setVelocity(new Vector(random.nextGaussian() * 0.001, 0.05, random.nextGaussian() * 0.001));
+
+        this.item = InventoryUtil.itemOrEmpty(item).clone();
+        int power = 0;
+        if (!InventoryUtil.isEmpty(item) && item.getItemMeta() instanceof FireworkMeta) {
+            this.metadata.set(MetadataIndex.FIREWORK_INFO, item.clone());
+            power = ((FireworkMeta) item.getItemMeta()).getPower();
+        }
+        lifeTime = calculateLifeTime(power);
+    }
+
+    @Override
+    public EntityType getType() {
+        return EntityType.FIREWORK;
+    }
+
+    @Override
+    public List<Message> createSpawnMessage() {
+        double x = location.getX();
+        double y = location.getY();
+        double z = location.getZ();
+
+        return Arrays.asList(
+            new SpawnObjectMessage(id, UUID.randomUUID(), SpawnObjectMessage.FIREWORK, x, y, z, 0, 0),
+            new EntityMetadataMessage(id, metadata.getEntryList())
+        );
+    }
+
+    @Override
+    public FireworkMeta getFireworkMeta() {
+        return ((FireworkMeta) item.getItemMeta()).clone();
+    }
+
+    @Override
+    public void setFireworkMeta(FireworkMeta fireworkMeta) {
+        if (fireworkMeta == null) {
+            return;
+        }
+
+        if (InventoryUtil.isEmpty(item) || !Material.FIREWORK.equals(item.getType())) {
+            item = new ItemStack(Material.FIREWORK);
+        }
+
+        this.item.setItemMeta(fireworkMeta.clone());
+        this.metadata.set(MetadataIndex.FIREWORK_INFO, item.clone());
+        this.lifeTime = calculateLifeTime(fireworkMeta.getPower());
+    }
+
+    private int calculateLifeTime(int power) {
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        return ((power + 1) * 10 + random.nextInt(0, 6) + random.nextInt(0, 7));
+    }
+
+    @Override
+    public void detonate() {
```

Comment:
I think this method should just make the firework explode immediately, unless I'm reading the Javadoc wrong.